docs/source/VUDRM-key-provision.md

```diff
@@ -0,0 +1,442 @@
+                return (Encoding.UTF8.GetString(plain, 0, decryptcount));
+            }
+            catch
+            {
+                return policy;
+            }
+        }
+
+        private string ComputeHash(string sharedSecret, string client, string encryptedMessage)
+        {
+            var preComputed = sharedSecret + client + encryptedMessage;
+            var algorithm = new SHA1Managed();
+            var bytesOf = Encoding.UTF8.GetBytes(preComputed);
+            var hashOf = algorithm.ComputeHash(bytesOf);
+            var result = new StringBuilder();
+            foreach (var b in hashOf)
+            {
+                result.Append(b.ToString("x2").ToLower());
+            }
+            return result.ToString();
+        }
+    }
+}
+```
+
+##### Ruby
+
+```ruby
+require 'base64'
+require 'crypt/rijndael'
+require 'digest'
+require 'openssl'
+
+client = "CLIENT_NAME"
+shared_secret = "SHARED_SECRET"
+
+encrypted_keys = "KEY_PROVIDER_RESPONSE"
+
+message, message_hash = encrypted_keys.split('|')
+
+pre_computed = client + message
+computed_hash = Digest::SHA1.new.hexdigest(shared_secret + pre_computed)
+
+if computed_hash == message_hash
+  decoded_message = Base64.strict_decode64(message)
+  decipher = OpenSSL::Cipher::Cipher.new('AES-128-ECB').decrypt
+  decipher.key = shared_secret[0..15]
+  unencrypted_keys = decipher.update(decoded_message) + decipher.final
+else
+  raise "Key provider response hash did not validate"
+end
+```
+
+##### GO
+
+```go
+package main
+
+import (
+	"crypto/aes"
+	"crypto/sha1"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+var (
+	Client = "CLIENT"
+	SharedSecret = "SHARED_SECRET"
+	KeyProviderResponse = "KEY_PROVIDER_RESPONSE"
+	PKCS5 = &pkcs5{}
+)
+
+type pkcs5 struct{}
+
+func main() {
+	s := strings.Split(KeyProviderResponse, "|")
+	encryptedMessage, hash := s[0], s[1]
+	generatedHash := generateKeyProviderHash(SharedSecret, Client, encryptedMessage)
+	if generatedHash == hash {
+		decryptedKeys := decrypt(encryptedMessage, SharedSecret)
+		fmt.Printf(decryptedKeys)
+	} else {
+		fmt.Printf("hash failed validation")
+	}
+}
+
+func generateKeyProviderHash(sharedSecret string, client string, encryptedMessage string) string {
+	v := fmt.Sprintf("%s%s%s", sharedSecret, client, encryptedMessage)
+	hasher := sha1.New()
+	hasher.Write([]byte(v))
+	return fmt.Sprintf("%x", hasher.Sum(nil))
+}
+
+func decrypt(encryptedData string, sharedSecret string) string {
+	source, _ := base64.StdEncoding.DecodeString(encryptedData)
+	key := []byte(sharedSecret)[0:16]
+
+	cipher, _ := aes.NewCipher([]byte(key))
+	buffer := make([]byte, len(source))
+	size := 16
+
+	for bs, be := 0, size; bs < len(source); bs, be = bs+size, be+size {
+		cipher.Decrypt(buffer[bs:be], source[bs:be])
+	}
+
+	plainBytes, _ := PKCS5.unPadding(buffer, 16)
+	return string(plainBytes)
+}
+
+func (p *pkcs5) unPadding(src []byte, blockSize int) ([]byte, error) {
+	srcLen := len(src)
+	paddingLen := int(src[srcLen-1])
```

Comment:
I still think this is a bug, but it might be solved by adding a simple `if srcLen % blockSize {` before going on to calculate paddingLen

Reply:
Why would this be a bug? The following if statement checks the padding size is ok.